server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -485,6 +486,36 @@ private static InternalAggregations reduceAggs(InternalAggregation.ReduceContext
             performFinalReduce ? aggReduceContextBuilder.forFinalReduction() : aggReduceContextBuilder.forPartialReduction());
     }
 
+    /**
+     * Checks that query results from all shards have consistent unsigned_long format.
+     * Sort queries on a field that has long type in one index, and unsigned_long in another index
+     * don't work correctly. Throw an error if this kind of sorting is detected.
+     * //TODO: instead of throwing error, find a way to sort long and unsigned_long together
+     */
+    private static void checkSameUnsignedLongSortFormat(Collection<? extends SearchPhaseResult> queryResults) {
+        boolean[] ulFormats = null;
+        boolean firstResult = true;
+        for (SearchPhaseResult entry : queryResults) {
+            DocValueFormat[] formats = entry.queryResult().sortValueFormats();
+            if (formats == null) return;
```

Comment:
Checking my understanding -- are all shards guaranteed to have the same number of sort formats (even if some sort fields are unmapped)?

Reply:
@jtibshirani Indeed, all shards are guaranteed to have the same number of sort formats. If a field is unmapped on a shard, we will get a shard failure, unless we specifically was [map an unmapped to something else](https://www.elastic.co/guide/en/elasticsearch/reference/current/sort-search-results.html#_ignoring_unmapped_fields).

Reply:
Also wondering if @jimczi is ok with this final check I've added in `SearchPhaseController`

Reply:
++, that makes sense to me. You can maybe rename the function validateMergeSortValueFormats or something along those lines ?